monai/networks/blocks/transformerblock.py

```diff
@@ -51,12 +64,161 @@ def __init__(
         if hidden_size % num_heads != 0:
             raise ValueError("hidden_size should be divisible by num_heads.")
 
+        if window_size > 0 and len(input_size) not in [2, 3]:
+            raise ValueError(
+                "If local window attention is used (window_size > 0), input_size should be specified: (h, w) or (h, w, d)"
+            )
+
         self.mlp = MLPBlock(hidden_size, mlp_dim, dropout_rate)
         self.norm1 = nn.LayerNorm(hidden_size)
         self.attn = SABlock(hidden_size, num_heads, dropout_rate, qkv_bias, save_attn)
         self.norm2 = nn.LayerNorm(hidden_size)
+        self.window_size = window_size
+        self.input_size = input_size
+
+    def forward(self, x: torch.Tensor):
+        """
+        Args:
+            x (Tensor): [b x (s_dim_1 * … * s_dim_n) x dim]
+        """
+        shortcut = x
+        x = self.norm1(x)
+        # Window partition
+        if self.window_size > 0:
+            if x.shape[1] != int(torch.prod(torch.tensor(self.input_size))):
+                raise ValueError(
+                    f"Input tensor spatial dimension {x.shape[1]} should be equal to {self.input_size} product"
+                )
+
+            if len(self.input_size) == 2:
+                x = rearrange(x, "b (h w) c -> b h w c", h=self.input_size[0], w=self.input_size[1])
```

Comment:
I would suggest moving this rearrange part to `window_partition` and 2d/3d into one `window_partition` func to make the code cleaner. After this PR, we may also consider integrating this [one](https://github.com/Project-MONAI/MONAI/commit/c61c6ac2d56af08fbbfb955324b8639e266a25db#diff-003ce93240016c5323702f64ea7b9f24d401ef98a8caf03763765cc958d176f6), which may make this `TransformerBlock` too complicated. What do you think?

Reply:
The main integration in that PR will be added the option of cross attention for conditioning. I'm not sure how the use of local window attention + cross attention should interact. Perhaps we would use the local windowing in the self-attention block, then not use it with the second, cross attention block? 

I think it should be OK to integrate the two, but I think the integration will be much easier/cleaner if we have a single `window_partition` func as @KumoLiu suggests.

Reply:
Thx for you comments @KumoLiu  and @marksgraham 
1. Yes you're right, i will clean the code following you suggestion
2. I also agree that the `TransformerBlock` can quickly become complex if we want to provide an exhaustive block to implement the networks presented in the literature (and those to come too). Regarding this what immediately comes to mind is defining different transformer blocks according to their subtleties (cross attention, self attention local window, and others...) rather than implement every variations in one block
3. Now considering the specifing issue of merging cross attention i don't see real issue.. local attention window was designed to be used in self-attention only (from my understanding). So cross attention should not interfere with the use of local windows for self attention. Also, i see that from https://github.com/Project-MONAI/MONAI/commit/c61c6ac2d56af08fbbfb955324b8639e266a25db#diff-003ce93240016c5323702f64ea7b9f24d401ef98a8caf03763765cc958d176f6 cross attention is defined using a self attention block, i would suggest to implement a cross attention block instead

Reply:
Agreed with point 2, at some point we should start splitting out into different transformer blocks that inherit from a base TransformerBlock class rather than trying to implement every possible variation in one.

And also agreed with 3 - I think this specific merge should be OK, either into one block or creating a CrossAttention block for the one above linked to.